framework/src/araui/core/ui_core.c

```diff
@@ -0,0 +1,835 @@
+{
+	ui_widget_body_t *parent = UI_NULL;
+
+	// If there is locked target, send all event to the target only
+	// else, send event to the focus widget and its hookers.
+
+	if (g_core.locked_target) {
+		g_core.locked_target->touch_cb(g_core.locked_target, touch_event, coord);
+	} else {
+		widget->touch_cb(widget, touch_event, coord);
+
+		// send touch event to the all hookers also
+		parent = widget->parent;
+		while (parent) {
+			if (parent->is_hooker && parent->touch_cb) {
+				parent->touch_cb(parent, touch_event, coord);
+			}
+			parent = parent->parent;
+		}
+	}
+}
+
+/**
+ * @brief Called when some widget is locked as an event target.
+ * But, the locked target doesn't get the event.
+ */
+static void _ui_cancel_touch_event_except_locked_target(ui_widget_body_t *widget)
+{
+	ui_coord_t null_coord = {0, };
+	ui_widget_body_t *parent;
+
+	// The focus widget should receive the cancel touch event
+	// whether the focus widget is a hooker or not.
+	if (widget->touch_cb && widget != g_core.locked_target) {
+		widget->touch_cb(widget, UI_TOUCH_EVENT_CANCEL, null_coord);
+	}
+
+	// Only hooker widgets inside the focus widget's parent list,
+	// should receive the touch cancel event.
+	parent = widget->parent;
+	while (parent) {
+		if (parent->is_hooker && parent->touch_cb && parent != g_core.locked_target) {
+			parent->touch_cb(parent, UI_TOUCH_EVENT_CANCEL, null_coord);
+		}
+		parent = parent->parent;
+	}
+}
+
+static void _ui_core_handle_touch_event(ui_touch_event_t touch_event, ui_coord_t coord)
+{
+	ui_view_body_t *view;
+	ui_widget_body_t *root;
+	ui_global_widget_body_t *global;
+
+	view = ui_view_get_current();
+	if (!view) {
+		UI_LOGE("error: active view is null!\n");
+		return;
+	}
+
+	root = view->root;
+
+	switch (touch_event) {
+		case UI_TOUCH_EVENT_DOWN: {
+			// If touch down, reset all infomations about focus and touch
+			ui_core_unlock_touch_event_target();
+
+			// If there is the visible global widget,
+			// a touch event will be delivered to only that global widget.
+			if (_ui_core_is_visible_global_widget()) {
+				global = (ui_global_widget_body_t *)g_global_widget_info[g_core.visible_event_type];
+
+				if (global->focus) {
+					_ui_cancel_touch_event_except_locked_target(global->focus);
+				}
+
+				global->focus = ui_widget_search_by_coord(g_global_widget_info[g_core.visible_event_type], coord);
+				if (global->focus && global->focus->touch_cb) {
+					_ui_deliver_touch_event(global->focus, touch_event, coord);
+				}
+				break;
+			}
+
+			// If a touch down event occurs at the area which global widget is set,
+			// a touch event will be delivered to only that global widget.
+			if (_ui_core_global_widget_touch_down(touch_event, coord)) {
+				break;
+			}
+
+			if (view->focus) {
+				_ui_cancel_touch_event_except_locked_target(view->focus);
+			}
+
+			view->focus = ui_widget_search_by_coord(root, coord);
+			if (view->focus) {
```

Comment:
Please assume that there's Circle in Rectangle. (circle is child of rectangle)
Circle doesn't have to handle touch event but rectangle need to handle it, also coordinate points center of circle.
It can handle this kind of situation?

Reply:
It can handle it.